File: control/recordings.mjs
// Read-only access to the captured recordings on disk.
//
// MediaMTX writes copy-only fMP4 to recordings/<cam>/<timestamp>.mp4. This lists
// them for the dashboard and resolves a single file for download — by camera +
// filename only (validated, joined under the recordings root) so a caller can
// never traverse outside it.
import {
  readdirSync, statSync, statfsSync, mkdirSync, writeFileSync, unlinkSync,
} from 'node:fs';
import { fileURLToPath } from 'node:url';
import { dirname, join, resolve, relative, isAbsolute } from 'node:path';
import { randomBytes } from 'node:crypto';

const ROOT = join(dirname(fileURLToPath(import.meta.url)), '..');
const RECORDINGS = join(ROOT, 'recordings');
const SAFE = /^[A-Za-z0-9._-]+$/;   // single path segment, no separators

function safeStat(p) {
  try {
    return statSync(p);
  } catch {
    return null;
  }
}

export function listRecordings() {
  const cams = [];
  const rootStat = safeStat(RECORDINGS);
  if (!rootStat || !rootStat.isDirectory()) return cams;
  for (const camName of readdirSync(RECORDINGS).sort()) {
    const camDir = join(RECORDINGS, camName);
    const cs = safeStat(camDir);
    if (!cs || !cs.isDirectory()) continue;
    const files = [];
    for (const fname of readdirSync(camDir).sort()) {
      const fpath = join(camDir, fname);
      const st = safeStat(fpath);
      if (st && st.isFile()) {
        files.push({ name: fname, sizeBytes: st.size, modified: st.mtimeMs / 1000 });
      }
    }
    cams.push({
      cam: camName,
      files,
      totalBytes: files.reduce((sum, x) => sum + x.sizeBytes, 0),
    });
  }
  return cams;
}

// Disk-readiness for recording: is the recordings folder writable, and how much
// free space is there. (Codec/audio/live are checked client-side from the
// MediaMTX paths list.)
export function preflight() {
  mkdirSync(RECORDINGS, { recursive: true });
  let writable = false;
  const probe = join(RECORDINGS, `.preflight-${randomBytes(6).toString('hex')}`);
  try {
    writeFileSync(probe, '');
    unlinkSync(probe);
    writable = true;
  } catch {
    writable = false;
  }
  let free = null;
  try {
    const fs = statfsSync(RECORDINGS);
    free = fs.bavail * fs.bsize;
  } catch {
    free = null;
  }
  return { recordingsWritable: writable, freeBytes: free, recordingsPath: RECORDINGS };
}

// Map (cam, name) to a file under the recordings root, or null if invalid.
export function resolveRecording(cam, name) {
  if (!(SAFE.test(cam || '') && SAFE.test(name || ''))) return null;
  if (cam === '..' || name === '..') return null;
  const root = resolve(RECORDINGS);
  const p = resolve(root, cam, name);
  const rel = relative(root, p);
  if (rel.startsWith('..') || isAbsolute(rel)) return null;   // reject anything outside the root
  const st = safeStat(p);
  return st && st.isFile() ? p : null;
}

File: control/app/recordings.py
"""Read-only access to the captured recordings on disk.

MediaMTX writes copy-only fMP4 to recordings/<cam>/<timestamp>.mp4. This lists
them for the dashboard and resolves a single file for download — by camera +
filename only (validated, joined under the recordings root) so a caller can
never traverse outside it.
"""
import re
import shutil
import tempfile
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
RECORDINGS = ROOT / "recordings"
_SAFE = re.compile(r"^[A-Za-z0-9._-]+$")   # single path segment, no separators


def list_recordings() -> list[dict]:
    cams: list[dict] = []
    if not RECORDINGS.is_dir():
        return cams
    for cam_dir in sorted(RECORDINGS.iterdir()):
        if not cam_dir.is_dir():
            continue
        files = []
        for f in sorted(cam_dir.iterdir()):
            if f.is_file():
                st = f.stat()
                files.append({"name": f.name, "sizeBytes": st.st_size, "modified": st.st_mtime})
        cams.append({"cam": cam_dir.name, "files": files,
                     "totalBytes": sum(x["sizeBytes"] for x in files)})
    return cams


def preflight() -> dict:
    """Disk-readiness for recording: is the recordings folder writable, and how
    much free space is there. (Codec/audio/live are checked client-side from the
    MediaMTX paths list.)"""
    RECORDINGS.mkdir(parents=True, exist_ok=True)
    writable = False
    try:
        with tempfile.NamedTemporaryFile(dir=RECORDINGS, prefix=".preflight-", delete=True):
            writable = True
    except Exception:
        writable = False
    try:
        free = shutil.disk_usage(RECORDINGS).free
    except Exception:
        free = None
    return {"recordingsWritable": writable, "freeBytes": free, "recordingsPath": str(RECORDINGS)}


def resolve_recording(cam: str, name: str) -> Path | None:
    """Map (cam, name) to a file under the recordings root, or None if invalid."""
    if not (_SAFE.match(cam or "") and _SAFE.match(name or "")):
        return None
    if cam == ".." or name == "..":
        return None
    root = RECORDINGS.resolve()
    p = (root / cam / name).resolve()
    try:
        p.relative_to(root)            # reject anything resolving outside the root
    except ValueError:
        return None
    return p if p.is_file() else None
